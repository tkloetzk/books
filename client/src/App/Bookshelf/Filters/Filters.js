import React, { Component } from 'react';
import FormControl from '@material-ui/core/FormControl';
import FormGroup from '@material-ui/core/FormGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { connect } from 'react-redux';
import Switch from '@material-ui/core/Switch';
import { withStyles } from '@material-ui/core/styles';
import { filterBookshelf } from '../../../store/bookshelf/bookshelfActions';

export class Filters extends Component {
  state = {
    filters: [
      {
        key: 'unread',
        value: false,
      },
      {
        key: 'owned',
        value: false,
      },
    ],
  };

  componentDidUpdate(prevProps, prevState) {
    const { filters } = this.state;
    const { filterBookshelf } = this.props;
    if (filters !== prevState.filters) {
      filterBookshelf(filters);
    }
  }

  handleOwned = name => {
    this.setState(prevState => ({
      filters: prevState.filters.map(filter =>
        filter.key === name
          ? Object.assign(filter, { value: !filter.value })
          : filter
      ),
    }));
  };

  render() {
    const { classes } = this.props;
    const { unread, owned } = this.state;
    const selectionControls = [
      {
        category: 'owned',
        checked: owned,
      },
      {
        category: 'unread',
        checked: unread,
      },
    ];

    return (
      <FormControl component="fieldset">
        <FormGroup row className={classes.formGroup}>
          {selectionControls.map(selection => {
            const { checked, category } = selection;
            return (
              <FormControlLabel
                control={
                  <Switch
                    color="primary"
                    checked={checked}
                    onChange={() => this.handleOwned(category)}
                    value={category}
                  />
                }
                label={category.toLocaleUpperCase()}
                key={category}
              />
            );
          })}
        </FormGroup>
      </FormControl>
    );
  }
}

const styles = {
  legend: {
    alignSelf: 'center',
    paddingRight: '10px',
  },
  formGroup: {
    justifyContent: 'center',
  },
};

const mapDispatchToProps = {
  filterBookshelf,
};

export default connect(
  null,
  mapDispatchToProps
)(withStyles(styles)(Filters));
