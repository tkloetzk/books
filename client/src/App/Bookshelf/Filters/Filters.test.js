import React from 'react';
import { mount, shallow } from 'enzyme';
import { Filters } from './Filters';

describe('Filters', () => {
  let props;
  let wrapper;
  let instance;
  beforeEach(() => {
    props = {
      classes: {},
      filterBookshelf: jest.fn(),
    };

    wrapper = shallow(<Filters {...props} />);
    instance = wrapper.instance();
  });

  it('renders with no function passed in', () => {
    wrapper = shallow(<Filters classes={{}} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with correct props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('updates state when handleOwned is called', () => {
    instance.handleOwned('owned');
    expect(instance.state.filters).toEqual([
      { key: 'unread', value: false },
      { key: 'owned', value: true },
    ]);

    instance.handleOwned('unread');
    expect(instance.state.filters).toEqual([
      { key: 'unread', value: true },
      { key: 'owned', value: true },
    ]);

    instance.handleOwned('owned');
    expect(instance.state.filters).toEqual([
      { key: 'unread', value: true },
      { key: 'owned', value: false },
    ]);

    instance.handleOwned('unread');
    expect(instance.state.filters).toEqual([
      { key: 'unread', value: false },
      { key: 'owned', value: false },
    ]);
  });

  it('calls filterBookshelf when filters are updated', () => {
    const prevState = [
      { key: 'unread', value: false },
      { key: 'owned', value: false },
    ];
    const filterState = [
      { key: 'unread', value: true },
      { key: 'owned', value: false },
    ];
    instance.state.filters = filterState;
    instance.componentDidUpdate(null, prevState);
    expect(instance.props.filterBookshelf).toHaveBeenCalled();
    expect(instance.props.filterBookshelf).toHaveBeenCalledWith(filterState);
  });
  it('does not call filterBookshelf when filters are not updated', () => {
    const filters = [
      { key: 'unread', value: false },
      { key: 'owned', value: false },
    ];
    instance.state.filters = filters;
    instance.componentDidUpdate(null, { filters });
    expect(instance.props.filterBookshelf).not.toHaveBeenCalled();
  });
});
