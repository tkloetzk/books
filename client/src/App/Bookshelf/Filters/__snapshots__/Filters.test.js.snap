// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters renders with correct props 1`] = `
<WithStyles(FormControl)
  component="fieldset"
>
  <WithStyles(FormGroup)
    row={true}
  >
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Switch)
          color="primary"
          onChange={[Function]}
          value="owned"
        />
      }
      key="owned"
      label="OWNED"
    />
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Switch)
          color="primary"
          onChange={[Function]}
          value="unread"
        />
      }
      key="unread"
      label="UNREAD"
    />
  </WithStyles(FormGroup)>
</WithStyles(FormControl)>
`;

exports[`Filters renders with no function passed in 1`] = `
<WithStyles(FormControl)
  component="fieldset"
>
  <WithStyles(FormGroup)
    row={true}
  >
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Switch)
          color="primary"
          onChange={[Function]}
          value="owned"
        />
      }
      key="owned"
      label="OWNED"
    />
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Switch)
          color="primary"
          onChange={[Function]}
          value="unread"
        />
      }
      key="unread"
      label="UNREAD"
    />
  </WithStyles(FormGroup)>
</WithStyles(FormControl)>
`;
